### DISCUSSION TITLE
Allow using different CapacityLimiters for Dependencies

### DISCUSSION THREAD

**Discussion by leo-b** [fastapi/fastapi#11886] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from contextlib import asynccontextmanager
import time
from fastapi import FastAPI, Depends
from anyio import CapacityLimiter
import anyio
from typing_extensions import Annotated

# proposed new feature
network_conn_limiter = CapacityLimiter(3)

@asynccontextmanager
async def lifespan(app: FastAPI):
    default_limiter = anyio.to_thread.current_default_thread_limiter()
    default_limiter.total_tokens = 2
    yield

app = FastAPI(lifespan=lifespan)

def slow_dependency():
    print(f"waiting for slow dependency")
    time.sleep(10)
    print(f"got slow dependency")
    return

@app.get("/")
def root():
    return {"message": "Hello World"}

@app.get("/slowdep")
def slowdep(dummy: Annotated[None, Depends(slow_dependency)]):
    return {"message": "slowdep"}

# proposed feature:
# @app.get("/slowdep2")
# def slowdep2(dummy: Annotated[None,
#                              Depends(slow_dependency,
#                                      limiter=network_conn_limiter)]):
#     return {"message": "slowdep2"}
```

### Description

For executing sync endpoints in a threadpool, FastAPIs routing code currently uses `run_in_threadpool()` which in turn uses `anyio.to_thread.run_sync()` with the _default_ `CapacityLimiter`.

FastAPI also runs sync dependencies via `run_in_threadpool()`, using the _default_ `CapacityLimiter` as well.

For some dependencies that access limited resources, using the same `CapacityLimiter` for running sync endpoints and dependencies could lead to _all_ sync endpoints being blocked because of just one dependency that consumes all threads of the common thread pool.

E.g. imagine some network device that is limited to e.g. 5 concurrent network connections. A sync dependency that establishes a new connection to such a device will yield the first five connections instantly but after that it will unnecessarily produce threads that just block waiting for a new connection. As soon as the default thread CapacityLimit of 40 threads is reached, even requests that don't even depend on those network connections will block because `routing.run_endpoint_function()` uses the same `CapacityLimiter` for sync endpoints.

If we could define a new `network_conn_limiter = CapacityLimiter(3)`
and use that limiter for just the one problematic dependency, starvation could be avoided because now other dependencies and routing are using different threads with different concurrency limits.

Cheers,
--leo

### Operating System

Linux

### FastAPI Version

0.111.1

### Pydantic Version

2.7.1

### Python Version

3.11.5

**Comment by leo-b** [comment]
PR https://github.com/tiangolo/fastapi/pull/11895 tries to implement this feature.

**Comment by yakubka** [comment]
AnyIO`s `CapacityLimiter` controls the maximum number of threads that can run concurrently in `run_sync_in_worker_thread` (what FastAPI uses internally for sync dependencies and handlers). The default limiter is shared globally.

To use a separate limiter for a specific dependency, call `anyio.to_thread.run_sync` directly with a custom limiter instead of relying on FastAPI`s automatic threadpool dispatch.

```python
import anyio
from fastapi import Depends

db_limiter = anyio.CapacityLimiter(5)  # max 5 concurrent DB threads

def blocking_db_query(id: int):
    return db.execute(f"SELECT * FROM items WHERE id = {id}")

async def get_item_from_db(item_id: int) -> dict:
    return await anyio.to_thread.run_sync(
        lambda: blocking_db_query(item_id),
        limiter=db_limiter,
    )

@app.get("/items/{item_id}")
async def get_item(item_id: int, result: dict = Depends(get_item_from_db)):
    return result
```

This gives the database dependency its own concurrency budget separate from the global thread pool. The same pattern works for any resource-constrained sync dependency.
